### PR TITLE
Move platform specific text selection behavior out of styled TextField classes

### LINKED ISSUE #129726: Separate text input behavior and UI: TextSelectionGestureDetectorBuilder
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

Anyone building their own text field based on EditableText may find they are missing certain behaviors that TextField and CupertinoTextField have built-in.  This is because some text editing behavior logic is located inside of TextField and CupertinoTextField.  Pulling this behavior code out of our UI libraries would solve the problem.

### Proposal

I propose to move text editing behavior logic out of the Cupertino and Material libraries and into `widgets`, specifically the logic in `TextSelectionGestureDetectorBuilder`'s subclasses.

The behavior of keyboard, tap, selection, and context menus during text editing should depend on the platform, not on the UI of the text field being used.  For example, a Material TextField being used in an iOS app should still have all of the same behaviors of any other text field on iOS: the iOS-style context menu should appear, iOS gestures should work, etc.

In order to do this in a DRY way, any text editing behavior logic should be defined at a lower level than the Cupertino and Material libraries.  This is currently not the case, as TextField and CupertinoTextField still have some behavior-related code.  Specifically, the logic in `_TextFieldSelectionGestureDetectorBuilder` in TextField and `_CupertinoTextFieldSelectionGestureDetectorBuilder` in CupertinoTextField should be moved into `TextSelectionGestureDetectorBuilder` in EditableText. Then switch statements on the platform should be used to differentiate platform behavior, not the UI library.

### Reference

This came up [in Discord](https://discord.com/channels/608014603317936148/608021234516754444/1123403298590957618) in relation to building a blank-canvas text field.  The general effort to separate text editing behavior from UI has been long running but never comprehensive (for example, the effort to move keyboard shortcuts out of Material/Cupertino in https://github.com/flutter/flutter/issues/75004).

### PR DESCRIPTION
**Pull request by Renzo-Olivares**
Fixes #129726

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 1]
```diff
@@ -2381,96 +2384,102 @@ class TextSelectionGestureDetectorBuilder {
...
+              }
+            } else if (((_positionWasOnSelectionExclusive(textPosition) && !previousSelection.isCollapsed)
+                       || (_positionWasOnSelectionInclusive(textPosition) && previousSelection.isCollapsed && isAffinityTheSame && !editableText.widget.readOnly))
+                       && renderEditable.hasFocus) {
+              editableText.toggleToolbar(false);
+            } else {
+              renderEditable.selectWordEdge(cause: SelectionChangedCause.tap);
+              if (previousSelection == editableText.textEditingValue.selection
+                  && renderEditable.hasFocus
+                  && !editableText.widget.readOnly) {
```
Comment (Renzo-Olivares) [inline packages/flutter/lib/src/widgets/text_selection.dart:2476]: should I use `renderEditable.readOnly` instead?

**Review by justinmc — APPROVED** [review]
Looks like a solid refactor, thanks for cleaning this up. Is there anything in Cupertino that can be similarly removed? I think originally iOS-specific behavior was put into Cupertino in places like _CupertinoTextFieldSelectionGestureDetectorBuilder, but really that platform-specific behavior should be in the widgets library.

File: packages/flutter/lib/src/widgets/text_selection.dart [hunk 2]
```diff
@@ -2381,96 +2384,103 @@ class TextSelectionGestureDetectorBuilder {
...
-                  editableText.toggleToolbar(false);
-                } else {
-                  editableText.hideToolbar(false);
-                }
+                editableText.hideToolbar(false);
               }
-          }
-      }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:2483]: I'll double check!

**Comment by Renzo-Olivares** [comment]
I didn't really see anything we could pull out of `CupertinoTextField`s implementation.

It mainly just has this, https://github.com/flutter/flutter/blob/3e4a57fa5000a15f2a133c1507522bb3ca293602/packages/flutter/lib/src/cupertino/text_field.dart#L111-L121
which seems `CupertinoTextField` specific since it handles the clear button behavior, so I think that's okay to stay there.

There's also https://github.com/flutter/flutter/blob/3e4a57fa5000a15f2a133c1507522bb3ca293602/packages/flutter/lib/src/cupertino/text_field.dart#L128
but i'm not quite sure what it's purpose is there for, and the git blame doesn't tell me much either. Removing it doesn't break any tests, do you have any thoughts?

File: packages/flutter/lib/src/widgets/text_selection.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (Renzo-Olivares) [reply packages/flutter/lib/src/widgets/text_selection.dart:2483]: I think this is right, the nested switch makes it look a little off.

**Comment by Renzo-Olivares** [comment]
Talked offline about https://github.com/flutter/flutter/pull/155774#issuecomment-2380289148 , will investigate https://github.com/flutter/flutter/blob/3e4a57fa5000a15f2a133c1507522bb3ca293602/packages/flutter/lib/src/cupertino/text_field.dart#L128 at a later time.

**Pull request merged**